### PR TITLE
Resolve by classloader and BCEL

### PR DESCRIPTION
**Pull request by suztomo**
ClassLoader helps to resolve methods with class hierarchy 
BCEL helps to find (package) private constructor or methods

### CODE REVIEW

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java [hunk 1]
```diff
@@ -211,8 +216,23 @@ public static void main(String[] arguments)
...
+        System.err.println("Jar file " + jarPath + " was not converted to URL: " + ex.getMessage());
+        return null;
+      }
+    }).filter(Objects::nonNull).toArray(URL[]::new);
+    URLClassLoader classLoaderFromJars = new URLClassLoader(jarFileUrls);
+
     for (FullyQualifiedMethodSignature methodReference : methodReferences) {
       String className = methodReference.getClassName();
+      if (className.startsWith("java.") || className.startsWith("sun.")
+          || className.startsWith("javax.")) {
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java]: javax classes are not necessarily in the JDK. They may need to be pulled in from outside.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java [hunk 2]
```diff
@@ -224,29 +244,99 @@ public static void main(String[] arguments)
...
-        availableMethodsInJars.addAll(availableMethodsOnClass);
-
-        if (!availableMethodsOnClass.contains(methodReference)) {
+        if (methodDefinitionExists(methodReference, classLoaderFromJars, repository)) {
+          availableMethodsInJars.add(methodReference);
+        } else {
           unresolvedMethods.add(methodReference);
         }
-      } catch (ClassNotFoundException ex) {
+      } catch (ClassNotFoundException|NoClassDefFoundError ex) {
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java]: spaces around |

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java (hunk 1 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java]: Removed "javax" from here

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java (hunk 2 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java]: Fixed

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java [hunk 3]
```diff
@@ -198,21 +201,37 @@ public static void main(String[] arguments)
    * @param methodReferences methods to search for within the jar files
    * @return list of methods that are not found in the jar files
    */
+  @VisibleForTesting
   static List<FullyQualifiedMethodSignature> findUnresolvedReferences(List<Path> jarFilePaths,
       List<FullyQualifiedMethodSignature> methodReferences) {
     List<FullyQualifiedMethodSignature> unresolvedMethods = new ArrayList<>();
 
-    // Creates chain of ClassPath items in the same order as jarFilePaths
+    // Creates chain of ClassPath items in the same order as jarFilePaths for BCEL API
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java]: classpath (i.e. one word so no internal capitalization)

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java [hunk 4]
```diff
@@ -198,21 +201,37 @@ public static void main(String[] arguments)
...
+
     Set<String> classesNotFound = new HashSet<>();
     Set<FullyQualifiedMethodSignature> availableMethodsInJars = new HashSet<>();
-    SyntheticRepository repository = SyntheticRepository.getInstance(classPath);
+
+    URL[] jarFileUrls = jarFilePaths.stream().map(jarPath -> {
+      try {
+        return jarPath.toUri().toURL();
+      } catch (MalformedURLException ex) {
+        System.err.println("Jar file " + jarPath + " was not converted to URL: " + ex.getMessage());
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java]: We may want to turn this into a logger, though maybe not in this PR.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java (hunk 3 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java]: Updated

**Pull request merged**